dev/src/reference.ts

```diff
@@ -2650,18 +2650,30 @@ export class Query<
             // If a non-transactional query failed, attempt to restart.
             // Transactional queries are retried via the transaction runner.
             if (!transactionId && !this._isPermanentRpcError(err, 'runQuery')) {
-              logger(
-                'Query._stream',
-                tag,
-                'Query failed with retryable stream error:',
-                err
-              );
-              // Enqueue a "no-op" write into the stream and resume the query
-              // once it is processed. This allows any enqueued results to be
-              // consumed before resuming the query so that the query resumption
-              // can start at the correct document.
-              stream.write(NOOP_MESSAGE, () => {
-                if (lastReceivedDocument) {
+              // Only streams that have received documents should be
+              // restarted to avoid rerunning long-running or high-cost
+              // queries that are not progressing to completion.
+              if (lastReceivedDocument) {
```

Comment:
I am very suspicious of this code, because I do not see `lastReceivedDocument` ever being reset to null during a retry.

I am also slightly concerned with accessing `lastReceivedDocument` outside of write callback, since we are dependent on stream pushing items through before this logic is evaluated. According to your tests this works, but it still smells like a race condition.

Reply:
> I am very suspicious of this code, because I do not see lastReceivedDocument ever being reset to null during a retry.

This was a bug that my tests did not catch. Good catch. Fixed and an additional test added.

> I am also slightly concerned with accessing lastReceivedDocument outside of write callback, since we are dependent on stream pushing items through before this logic is evaluated. According to your tests this works, but it still smells like a race condition.

I also was initially. Here's why I decided on this. The noop write only makes sense if we are resuming the query, so we do need to check `lastReceivedDocument` outside of the write callback. I added an assertion to validate that `lastReceivedDocument` did not change and I added a comment regarding why it should never change.

```
// `backendStream` was unpiped at the beginning of the handler
// for `backendStream.on('error', ...)`, so we don't expect
// `lastReceivedDocument` to be updated during the noop write,
// but a check here is included for this unexpected state.
assert(
  lastReceivedDocument !== null,
  'lastReceivedDocument is unset'
);
```

Reply:
This still makes me uncomfortable because of buffering:
https://nodejs.org/api/stream.html#buffering

If you add `execSync('sleep 1');` to Transform above (in an attempt to cause buffering) does it still work?